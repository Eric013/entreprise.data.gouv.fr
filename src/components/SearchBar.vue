<template>
  <div class="row justify-content-center">
    <div class="col-md-6">
		<h2 class="search-title">{{ searchName}}</h2>
      <div id="custom-search-input">
        <div class="input-group col-md-12">
          <input type="text" class="form-control input-lg" placeholder="Recherche" />
          <span class="input-group-btn">
            <button class="btn btn-info btn-lg" type="button">
              <svg class="icon icon-search"><use xlink:href="#icon-search"></use></svg>
            </button>
          </span>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'SearchBar',
  props: ['searchName']
}
</script>

<style scoped>
  #custom-search-input{
      padding: 3px;
      border: solid 1px #E4E4E4;
      border-radius: 6px;
      background-color: #fff;
  }

  #custom-search-input input{
      border: 0;
      box-shadow: none;
  }

  #custom-search-input button{
      margin: 2px 0 0 0;
      background: none;
      box-shadow: none;
      border: 0;
      color: #666666;
      padding: 0 8px 0 10px;
      border-left: solid 1px #ccc;
  }

  #custom-search-input button:hover{
      border: 0;
      box-shadow: none;
      border-left: solid 1px #ccc;
  }

  #custom-search-input .glyphicon-search{
      font-size: 23px;
  }
</style>
