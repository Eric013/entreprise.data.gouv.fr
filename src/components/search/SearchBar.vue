<template>
  <div class="form__group">
    <!-- <div class="form__select">
      <span>Je recherche une :</span>
      <input type="radio" name="search_type" value="sirene"> <span>Association</span>
      <input type="radio" name="search_type" value="rna"> <span>Entreprise</span>
    </div> -->
    <input type="text" name="search" placeholder="Nom, SIREN, SIRET, adresse..." v-model="fullText"
      @keydown.down="suggestDown"
      @keydown.up.prevent="suggestUp"
      @keydown.esc="suggestReset"
      @keydown.enter="requestSearchIfNotEmpty"/>
    <button class="overlay-button">
      <svg class="icon icon-search"><use xlink:href="#icon-search"></use></svg>
    </button>

    <ul v-show="suggestions">
      <li class="suggestion__box"
          v-for="(suggestion, index) in suggestions"
          :key="index"
          v-bind:class="{'active': suggestActive(index)}"
          @click="suggestSelectAndEnter(index)">
        <span>{{ suggestion | capitalize | removeExtraChars}}</span>
      </li>
      <!-- Filling with hidden divs so search bar will always be same size -->
      <li class="hidden suggestion__box" :key="index" v-for="index in suggestionNumberToMax"></li>
    </ul>
  </div>
</template>

<script>
import Filters from '@/components/mixins/filters.js'
import SuggestionsHelpers from '@/components/mixins/suggestionsHelpers.js'
import RegExps from '@/components/mixins/regExps.js'

export default {
  name: 'SearchBar',
  props: ['searchName'],
  computed: {
    fullText: {
      get: function () {
        return this.$store.state.search.storedFullText
      },
      set: function (fullText) {
        this.$store.commit('setFullText', fullText)
        if (!fullText) {
          this.$store.dispatch('goToClearedHomePage')
        }
        if (String(fullText).length >= 3) {
          this.resetIndexSuggestion()
          this.$store.commit('setPage', 1)
          this.$store.commit('setQuerySuggestions', fullText)
          this.$store.dispatch('executeSearchSuggestions')
        }
      }
    },
    isSearchNotEmpty () {
      return this.$store.state.search.storedFullText !== ''
    },
  },
  methods: {
    requestSearchIfNotEmpty: function() {
      if (this.isSearchNotEmpty) {
        this.requestSearch()
      }
    },
    requestSearch: function () {
      const natureSearchId = this.analyzeSearchId(this.fullText)

      if (natureSearchId) {
        this.fullText = this.removeSeparators(this.fullText)
        this.$router.push({ path: `/etablissement/${this.fullText}` })
      } else {
        this.requestFullTextSearch()
      }
      this.$store.commit('clearResults')
    },
    requestFullTextSearch: function () {
      const currentSuggestion = this.currentSuggestion()
      if (currentSuggestion) {
        this.$store.commit('setFullText', currentSuggestion)
      } else {
        const fullTextNoDiacritics = this.removeDiacritics(this.fullText)
        this.$store.commit('setFullText', fullTextNoDiacritics)
      }
      this.$store.dispatch('requestSearchFullText')
      this.suggestCount = -1
    }
  },
  mixins: [Filters, SuggestionsHelpers, RegExps]
}
</script>

<style lang="scss" scoped>
  .form__group {
    width: 35em;
    max-width: 100%;
  }

  .form__select {
    margin-bottom: 5px;
  }

  ul {
    list-style-type: none;
    padding: 0;
    margin: 0;
  }

  .suggestion__box {
      min-height: 2.7em;
      outline: none;
      padding: 8px 14px;
      font: inherit;
      line-height: 1.6;
      color: $color-black;
      border-radius: 3px;
      -webkit-box-sizing: border-box;
              box-sizing: border-box;
      border: 1px solid $color-grey-blue;
      background: $color-white;
      vertical-align: middle;
      position: relative;
  }

  // TODO: is this called ?
  .hidden {
    visibility: hidden;
  }

  .active {
    background-color: $color-light-blue;
    color: $color-white;
  }
</style>
