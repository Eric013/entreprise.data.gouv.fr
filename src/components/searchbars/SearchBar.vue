<template>
  <div class="row justify-content-center">
    <div class="col-6">
		<h2 class="search-title">{{ searchName }}</h2>
      <div id="custom-search-input">
        <div class="input-group col-12">
          <input type="text" class="form-control input-lg" placeholder="Recherche" v-model="fullText"/>
          <span class="input-group-btn">
            <button class="btn btn-info btn-lg" type="button">
              <svg class="icon icon-search"><use xlink:href="#icon-search"></use></svg>
            </button>
          </span>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import store from '../../store/store.js'

export default {
  name: 'SearchBar',
  props: ['searchName'],
  data () {
    return {
      fullText: ''
    }
  },
  watch: {
    // whenever fullText changes, this function will run
    /* eslint-disable no-undef */
    fullText: function (setFullText) {
      store.commit('setFullText', this.fullText)
      store.commit('executeSearch')
    }
  }
}
</script>

<style scoped>
  #custom-search-input{
      padding: 3px;
      border: solid 1px #E4E4E4;
      border-radius: 6px;
      background-color: #fff;
  }

  #custom-search-input input{
      border: 0;
      box-shadow: none;
  }

  #custom-search-input button{
      margin: 2px 0 0 0;
      background: none;
      box-shadow: none;
      border: 0;
      color: #666666;
      padding: 0 8px 0 10px;
      border-left: solid 1px #ccc;
  }

  #custom-search-input button:hover{
      border: 0;
      box-shadow: none;
      border-left: solid 1px #ccc;
  }

  #custom-search-input {
      font-size: 23px;
  }

  .icon-search {
    font-size: 23px;
  }
</style>
