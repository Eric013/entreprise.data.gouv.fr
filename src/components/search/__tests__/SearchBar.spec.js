import { createLocalVue, shallow, mount } from '@vue/test-utils'
import Vuex from 'vuex'
import SearchBar from '@/components/search/SearchBar'
import { __createMocks as createStoreMocks } from '@/store/index.js'

const localVue = createLocalVue()
localVue.use(Vuex)
jest.mock('@/store/index.js')

describe('SearchBar.vue', () => {
  let storeMocks
  let wrapperSearchBar
  const $router = {
    push: jest.fn()
  }

  beforeEach(() => {
    storeMocks = createStoreMocks()
    
    wrapperSearchBar = shallow(SearchBar, {
      localVue,
      store: storeMocks.store,
      mocks: { $router }
    })
  })

  // Tests for Suggestions
  test('When searching under 3 chars, a request for suggestions should NOT be made', () => {
    wrapperSearchBar.setData({ fullText: 'ma' })
    expect(storeMocks.actions.executeSearchSuggestions.mock.calls).toHaveLength(0)
  }),

  test('When searching over/equal 3 chars, a request for suggestions should be made', () => {
    wrapperSearchBar.setData({ fullText: 'maRequete' })
    expect(storeMocks.actions.executeSearchSuggestions).toHaveBeenCalled()
  }),

  test('If the request for suggestions return >= 5 results, 5 suggestions should be displayed', () => {
    wrapperSearchBar.vm.$store.state.suggestions.storedSuggestions = ['Suggest1', 'Suggest2', 'Suggest3', 'Suggest4', 'Suggest5', 'Suggest6']
    // Need to update twice before suggestions are rendered
    wrapperSearchBar.update()
    wrapperSearchBar.update()
    expect(wrapperSearchBar.findAll('li' + '.suggestion__box')).toHaveLength(5)
    expect(wrapperSearchBar.findAll('li' + '.suggestion__box' + '.hidden')).toHaveLength(0)
  }),

  test('If the request for suggestions return 3 results, 3 suggestions should be displayed + 2 hidden', () => {
    storeMocks.state.suggestions.storedSuggestions = ['Suggest1', 'Suggest2', 'Suggest3']
    wrapperSearchBar.update()
    expect(wrapperSearchBar.findAll('li' + '.suggestion__box')).toHaveLength(5)
    expect(wrapperSearchBar.findAll('li' + '.suggestion__box' + '.hidden')).toHaveLength(2)
  }),

  test('Pressing escape key reset storedSuggestions', () => {
    storeMocks.state.suggestions.storedSuggestions = ['Suggest1', 'Suggest2', 'Suggest3']
    wrapperSearchBar.update()
    wrapperSearchBar.find('input').trigger('keydown.esc')
    expect(storeMocks.mutations.setStoredSuggestions).toHaveBeenCalledWith(storeMocks.state, '')
  }),

  test("When storedSuggestion is '', no suggestions are displayed" , () => {
    storeMocks.state.suggestions.storedSuggestions = ''
    wrapperSearchBar.update()
    const suggestionsWrappers = wrapperSearchBar.findAll('li' + '.suggestion__box').wrappers
    // Suggestions are rendered in virtual DOM but not attached to document (displayed)
    Array.prototype.forEach.call(suggestionsWrappers, (element) =>
      expect(element.options.attachedToDocument).toBeFalsy())
  }),

  test('I can highlight a suggestion by using keydown', () => {
    storeMocks.state.suggestions.storedSuggestions = ['Suggest1', 'Suggest2', 'Suggest3']
    wrapperSearchBar.update()
    // We go down 3 times to get Suggest3
    wrapperSearchBar.find('input').trigger('keydown.down')
    wrapperSearchBar.find('input').trigger('keydown.down')
    wrapperSearchBar.find('input').trigger('keydown.down')
    expect(wrapperSearchBar.findAll('.active')).toHaveLength(1)
    expect(wrapperSearchBar.find('.active').text()).toBe('Suggest3')
  }),

  test('When highlighting suggestions down, I dont exceed the last suggestion', () => {
    storeMocks.state.suggestions.storedSuggestions = ['Suggest1', 'Suggest2', 'Suggest3']
    // We go down 3 times to get Suggest3
    wrapperSearchBar.find('input').trigger('keydown.down')
    wrapperSearchBar.find('input').trigger('keydown.down')
    wrapperSearchBar.find('input').trigger('keydown.down')
    wrapperSearchBar.find('input').trigger('keydown.down')
    expect(wrapperSearchBar.findAll('.active')).toHaveLength(1)
    expect(wrapperSearchBar.find('.active').text()).toBe('Suggest3')
  }),

  test('I can highlight suggestions up too', () => {
    storeMocks.state.suggestions.storedSuggestions = ['Suggest1', 'Suggest2', 'Suggest3']
    // We go down 3 times to get Suggest3
    wrapperSearchBar.find('input').trigger('keydown.down')
    wrapperSearchBar.find('input').trigger('keydown.up')
    wrapperSearchBar.find('input').trigger('keydown.down')
    expect(wrapperSearchBar.findAll('.active')).toHaveLength(1)
    expect(wrapperSearchBar.find('.active').text()).toBe('Suggest1')
  }),

  test('When highlighting suggestions up, I dont go too far up', () => {
    storeMocks.state.suggestions.storedSuggestions = ['Suggest1', 'Suggest2', 'Suggest3']
    // We go down 3 times to get Suggest3
    wrapperSearchBar.find('input').trigger('keydown.down')
    wrapperSearchBar.find('input').trigger('keydown.up')
    wrapperSearchBar.find('input').trigger('keydown.up')
    wrapperSearchBar.find('input').trigger('keydown.up')
    wrapperSearchBar.find('input').trigger('keydown.down')
    expect(wrapperSearchBar.findAll('.active')).toHaveLength(1)
    expect(wrapperSearchBar.find('.active').text()).toBe('Suggest1')
  }),

  test('I can press enter to launch a search on the highlighted suggestion', () => {
    storeMocks.state.suggestions.storedSuggestions = ['Suggest1', 'Suggest2', 'Suggest3']
    wrapperSearchBar.vm.requestSearch = jest.fn()
    // wrapperSearchBar.vm.$store._mutations.setFullText = jest.fn()
    // We go down 3 times to get Suggest3
    wrapperSearchBar.find('input').trigger('keydown.down')
    wrapperSearchBar.find('input').trigger('keydown.down')
    wrapperSearchBar.find('input').trigger('keydown.down')
    wrapperSearchBar.find('input').trigger('keydown.enter')
    // requestSearch use currentSuggestion as fullText
    expect(wrapperSearchBar.vm.currentSuggestion()).toBe('Suggest3')
    expect(wrapperSearchBar.vm.requestSearch.mock.calls).toHaveLength(1)
  }),

  test('I can click on a suggestion to launch a search', () => {
    storeMocks.state.suggestions.storedSuggestions = ['Suggest1', 'Suggest2', 'Suggest3']
    wrapperSearchBar.vm.requestSearch = jest.fn()  
    wrapperSearchBar.find('li:nth-child(2)').trigger('click')
    expect(wrapperSearchBar.vm.currentSuggestion()).toBe('Suggest2')
    expect(wrapperSearchBar.vm.requestSearch.mock.calls).toHaveLength(1)
  })
  // End Tests for Suggestions

  // Start Tests for Methods

  test('Method requestSirenSearch dispatch executeSearchBySiren with fullText', () => {
    wrapperSearchBar.vm.requestSirenSearch()
    expect(storeMocks.actions.executeSearchBySiren.mock.calls[0][1]).toBe('mock-storedFullText')
  })

  test('Method requestSirenSearch push router to siegeSiret (if requestSirenSearch was successful)', () => {
    wrapperSearchBar.vm.requestSirenSearch()
    expect(wrapperSearchBar.vm.$router.push).toHaveBeenCalledWith({ path: `/entreprise/mock-storedSirenSiegeSiret` })
  })

  // TODO
  test('Method requestSirenSearch set response at notFound (if requestSirenSearch was not successful)')

  test('Method requestSiretSearch set siret as fullText then push router to fullText', () => {
    wrapperSearchBar.vm.requestSiretSearch()
    expect(storeMocks.mutations.setSiret).toHaveBeenCalledWith(storeMocks.state, 'mock-storedFullText')
    expect(wrapperSearchBar.vm.$router.push).toHaveBeenCalledWith({ path: `/entreprise/mock-storedFullText` })
  })

  test('Method requestFullTextSearch search suggestion if it is set', () => {
    wrapperSearchBar.vm.currentSuggestion = jest.fn().mockImplementation(() => { return 'mock-suggest' })
    wrapperSearchBar.vm.requestFullTextSearch()
    expect(wrapperSearchBar.vm.currentSuggestion).toHaveBeenCalled()
    expect(storeMocks.mutations.setFullText).toHaveBeenCalledWith(storeMocks.state, 'mock-suggest')
  })

  test('Method requestFullTextSearch search fullText if suggestion not set', () => {
    wrapperSearchBar.vm.currentSuggestion = jest.fn().mockImplementation(() => { return null })
    wrapperSearchBar.vm.requestFullTextSearch()
    expect(storeMocks.mutations.setFullText).toHaveBeenCalledWith(storeMocks.state, 'mock-storedFullText')
  })

  test('Method requestFullTextSearch remove diacritics', () => {
    wrapperSearchBar.vm.removeDiacritics = jest.fn().mockImplementation(() => { return 'removed-diacritics' })
    wrapperSearchBar.vm.requestFullTextSearch()
    expect(storeMocks.mutations.setFullText).toHaveBeenCalledWith(storeMocks.state, 'removed-diacritics')
  })

  test('Method requestFullTextSearch request a search', () => {
    wrapperSearchBar.vm.requestFullTextSearch()
    expect(storeMocks.actions.requestSearch).toHaveBeenCalled()
  })

  test('Method requestFullTextSearch sets suggestCount at -1', () => {
    wrapperSearchBar.vm.requestFullTextSearch()
    expect(wrapperSearchBar.vm.suggestCount).toBe(-1)
  })

  test('Method requestSearch remove separators and requestSiretSearch if fullText is siret', () => {
    wrapperSearchBar.vm.removeSeparators = jest.fn()
    wrapperSearchBar.vm.requestSiretSearch = jest.fn()
    wrapperSearchBar.vm.isSiret = jest.fn().mockReturnValue(true)
    wrapperSearchBar.vm.isSiren = jest.fn().mockReturnValue(false)
    wrapperSearchBar.vm.requestSearch()
    expect(wrapperSearchBar.vm.removeSeparators).toHaveBeenCalledWith('mock-storedFullText')
    expect(wrapperSearchBar.vm.requestSiretSearch).toHaveBeenCalled()
  })

  test('Method requestSearch remove separators if fullText is siren', () => {
    wrapperSearchBar.vm.removeSeparators = jest.fn()
    wrapperSearchBar.vm.requestSirenSearch = jest.fn()
    wrapperSearchBar.vm.isSiret = jest.fn().mockReturnValue(false)
    wrapperSearchBar.vm.isSiren = jest.fn().mockReturnValue(true)
    wrapperSearchBar.vm.requestSearch()
    expect(wrapperSearchBar.vm.removeSeparators).toHaveBeenCalledWith('mock-storedFullText')
    expect(wrapperSearchBar.vm.requestSirenSearch).toHaveBeenCalled()
  })

  test('Method requestSearch launch requestFullTextSearch if fullText not siret nor siren', () => {
    wrapperSearchBar.vm.requestFullTextSearch = jest.fn()
    wrapperSearchBar.vm.isSiret = jest.fn().mockReturnValue(false)
    wrapperSearchBar.vm.isSiren = jest.fn().mockReturnValue(false)
    wrapperSearchBar.vm.requestSearch()
    expect(wrapperSearchBar.vm.requestFullTextSearch).toHaveBeenCalled()
  })

  test('Method requestSearchIfNotEmpty doesnt search if search is empty', () => {
    storeMocks.store.state.search.storedFullText = ''
    wrapperSearchBar.vm.requestSearch = jest.fn()
    wrapperSearchBar.vm.requestSearchIfNotEmpty()
    expect(wrapperSearchBar.vm.requestSearch).toHaveBeenCalledTimes(0)
  })

  // Computed values
  
  test('Computed value isSearchNotEmpty check if search is empty', () =>{
    storeMocks.store.state.search.storedFullText = ''
    expect(wrapperSearchBar.vm.isSearchNotEmpty).toBeFalsy()
    storeMocks.store.state.search.storedFullText = 'placeholder'
    expect(wrapperSearchBar.vm.isSearchNotEmpty).toBeTruthy()
  })

})


describe('SearchBar.vue snapshot', () => {
  const storeMocks = createStoreMocks()

  const wrapperSearchBar = mount(SearchBar, {
    localVue,
    store: storeMocks.store
  })

  test('Snapshot test', () => {
    expect(wrapperSearchBar.vm.$el).toMatchSnapshot()
  })
})
