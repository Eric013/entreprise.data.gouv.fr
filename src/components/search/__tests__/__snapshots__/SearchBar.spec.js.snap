// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar.vue snapshot Snapshot test 1`] = `
<div
  class="form__group"
>
  <input
    name="search"
    placeholder="Nom, SIREN, SIRET, adresse..."
    type="text"
  />
   
  <button
    class="overlay-button"
  >
    <svg
      class="icon icon-search"
    >
      <use
        xlink:href="#icon-search"
      />
    </svg>
  </button>
   
  <ul>
     
    <li
      class="hidden suggestion__box"
    />
    <li
      class="hidden suggestion__box"
    />
    <li
      class="hidden suggestion__box"
    />
    <li
      class="hidden suggestion__box"
    />
    <li
      class="hidden suggestion__box"
    />
  </ul>
</div>
`;
